### ISSUE TITLE
Variant prices reverting to original value

### ISSUE THREAD

**Issue by mkllnk** [openfoodfoundation/openfoodnetwork#2151]
A user in Australia reported price changes that they didn't do. The user is the only manager. They said it looks like the price they originally entered for the product when they started using the OFN.

I'm still gathering information about this bug and will fill in this issue as I go.

## Description

Affected enterprise id: 287 - producer
Affected product id: 1395, last updated 2018-03-17 00:36:38 UTC which is 11:36 AEDT.

I downloaded a database backup from this morning and the variant had indeed a price of $8 instead of $19.50. The product had last been changed at 2018-02-19 01:55:59 UTC.

The user reports that order on 1 March were correct. Orders after that date had the wrong price. Pull requests merged around that time:

- https://github.com/openfoodfoundation/openfoodnetwork/pull/2087
- https://github.com/openfoodfoundation/openfoodnetwork/pull/2089
- https://github.com/openfoodfoundation/openfoodnetwork/pull/2088
- https://github.com/openfoodfoundation/openfoodnetwork/pull/2086
- https://github.com/openfoodfoundation/openfoodnetwork/pull/2069

We moved production servers around that time with the final database switch on Saturday 3 March.

Looking at the old database, I found that the affected variant has two prices in the database:
```
openfoodweb_production=> select * from spree_prices where variant_id = 6385;
  id  | variant_id | amount | currency 
------+------------+--------+----------
 8747 |       6385 |  19.50 | AUD
 8746 |       6385 |   8.00 | AUD
(2 rows)
```

This is true for all variants. After the user updated the price to correct it, both prices were the same.

The query generated by ActiveRecord to get the price of a variant is:
```sql
SELECT "spree_prices".* FROM "spree_prices" WHERE "spree_prices"."variant_id" = 6385 AND "spree_prices"."currency" = 'AUD' LIMIT 1
```

That tells me that there should be only one price and it's random which one is retrieved from the database. Random meaning, it depends in which order the database stored the information. This could have changed by dumping and re-importing the database or it could depend on the Postgresql version.

My theory is that moving to a new server, caused this random selection of the two prices to change.

Duplicate price rows have been on the Aus production server in December already. That means that our server change did not cause the duplication, but made it visible.

Affected databases:

- Aus production
- UK production

Unaffected databases:

- Aus staging1
- Canada production
- Katuma

## Expected Behavior

Price should not have changed.

## Actual Behavior

Price reverted to original.

## Steps to Reproduce

1. Visit https://openfoodnetwork.org.au/admin/products/bulk_edit
2. Search product Dashi Bonito
3. Check price

Alternatively:

```
Spree::Product.find(1395).variants.first.price
```

## Context

The person managing the products is updating them monthly according to supplier updates.

## Severity

Bad. Possible price changes from $19.50 to $8.

## Your Environment

Aus production is currently running ada0fa50fc08f057bab4e353d451070438e5dc3b deployed on Wednesday 14 March 2018.

## Possible Fix

Monitoring prices and changing them back at the moment.

**Comment by myriamboure** [comment]
Can't it be a user mistake? Maybe they are a couple of managers and someone did a mistake? This happened a couple of time in France those kine of "bugs" that where not :-o Maybe that's not the case here thought ;-)

**Comment by mkllnk** [comment]
There is only one manager.

**Comment by mkllnk** [comment]
If @oeoeaio wants to take over from here. Some SQL I used to identify affected orders:
```sql
select distinct(e.name, o.number) from enterprises e, spree_variants v, spree_orders o where o.distributor_id=e.id and o.id in (select order_id from spree_line_items where order_id in (select id from spree_orders where completed_at > '2018-03-02') and variant_id in (select p1.variant_id from spree_prices p1 join spree_prices p2 on p1.variant_id=p2.variant_id and p1.id!=p2.id where p1.amount!=p2.amount and p1.currency = p2.currency))

select distinct e.name, p.name product, pr.amount other_price from spree_products p join spree_variants v on p.id=v.product_id and v.is_master='f' join spree_line_items i on i.variant_id=v.id join spree_orders o on i.order_id=o.id join enterprises e on o.distributor_id=e.id join spree_prices pr on v.id=pr.variant_id and i.price!=pr.amount and i.currency=pr.currency where order_id in (select id from spree_orders where completed_at > '2018-03-02') and i.variant_id in (select p1.variant_id from spree_prices p1 join spree_prices p2 on p1.variant_id=p2.variant_id and p1.id!=p2.id where p1.amount!=p2.amount and p1.currency = p2.currency) order by e.name;
```

**Comment by oeoeaio** [comment]
@Matt-Yorkley @lin-d-hop You may need to deal with this, especially if you are considering migrating to a new production server at any point. Sounds like it is not an issue for anyone else?

The summary of the issue is that there were a bunch of variants that somehow had multiple prices in AUD on our old production server. This was not a problem because the way prices are loaded appears to be pseudo-deterministic: the same record was always returned when we asked for `variant.price`. In our case, the price with the HIGHER `id` was always returned, which is probably something to do with how the database asks for records?

When we migrated to the new production server, the opposite started happening: for variants with multiple prices in AUD, the price with the LOWER id seems to be returned consistently. This caused the issue described above: "old" prices started popping up across the system for a bunch of variants.

We have hopefully resolved this issue on Aus production by running the script in #2152, which is more complicated than you will need because we had to work out a way of keeping prices that people had altered since the migration to the new server. Hopefully you can just work out whether your database is currently using the HIGHER or LOWER ids for prices and then simply delete the one that is not used for each variants with duplicates.

Let me know if you need a hand with anything, or want @mkllnk or I to have a look at something.

**Comment by oeoeaio** [comment]
@sauloperez Only affects variants that were created before about the end of May 2015, so I suspect it was some migration that did something bad. I suspect it had something to do with `db/migrate/20141003060622_add_standard_variant_to_products.rb`,but I haven't worked out what went wrong. I guess the only instances that ever had variants before that date were Aus and UK, so they are the only ones affected.

**Comment by Matt-Yorkley** [comment]
Resolved by #2152
